src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs

```diff
@@ -72,18 +74,66 @@ internal partial class CSharpSelectionValidator : SelectionValidator
                 }
             }
 
+            // Warn if local functions are in selection since data flow analysis
+            // cannot correctly analyze them
+            // https://github.com/dotnet/roslyn/issues/14214
+            if (SpanContainsLocalFunction(selectionInfo.FinalSpan, model, root))
+            {
+                selectionInfo = selectionInfo.WithStatus(s => s.With(
+                    OperationStatusFlag.Succeeded | OperationStatusFlag.BestEffort,
+                    CSharpFeaturesResources.Local_func_in_span));
+                var commonRoot = selectionInfo.CommonRootFromOriginalSpan;
+                var annotated = commonRoot.WithAdditionalAnnotations(
+                        WarningAnnotation.Create(CSharpFeaturesResources.Local_func_in_span));
+                doc = await doc.WithSyntaxRootAsync(
+                    root.ReplaceNode(commonRoot, annotated),
+                    cancellationToken).ConfigureAwait(false);
+                selectionInfo.FirstTokenInOriginalSpan = doc.Root.FindToken(selectionInfo.FirstTokenInOriginalSpan.SpanStart);
+                selectionInfo.LastTokenInOriginalSpan = doc.Root.FindToken(selectionInfo.LastTokenInOriginalSpan.SpanStart);
+                selectionInfo.FirstTokenInFinalSpan = doc.Root.FindToken(selectionInfo.FirstTokenInFinalSpan.SpanStart);
+                selectionInfo.LastTokenInFinalSpan = doc.Root.FindToken(selectionInfo.LastTokenInFinalSpan.SpanStart);
+            }
+
             return await CSharpSelectionResult.CreateAsync(
                 selectionInfo.Status,
                 selectionInfo.OriginalSpan,
                 selectionInfo.FinalSpan,
                 this.Options,
                 selectionInfo.SelectionInExpression,
-                this.SemanticDocument,
+                doc,
                 selectionInfo.FirstTokenInFinalSpan,
                 selectionInfo.LastTokenInFinalSpan,
                 cancellationToken).ConfigureAwait(false);
         }
 
+        private static bool SpanContainsLocalFunction(TextSpan finalSpan, SemanticModel model, SyntaxNode root)
+        {
+            var nodes = root.DescendantNodes().Where(n => finalSpan.Contains(n.Span));
```

Comment:
Note: DescendantNodes can take a span of interest (and the walk will be more efficient).  You will get all nodes that intersect the span.  You can then iterate over that, ignoring nodes that are not contained in finalSpan.